subprojects/core/src/main/java/org/gradle/internal/build/ResilientBuildToolingModelController.java

```diff
@@ -16,68 +16,149 @@
+            boolean isOwnerBuildSuccessfullyConfigured = target.getOwner().isBuildConfigured();
+            return new ResilientToolingModelBuilder(builder, configurationResult, isConfigureOnDemand, isOwnerBuildSuccessfullyConfigured, failureFactory, modelName);
+        }
+    }
+
+    private static class ResilientToolingModelBuilder implements ToolingModelBuilderLookup.Builder {
+
+        private final ToolingModelBuilderLookup.Builder delegate;
+        private final ConfigurationResult configurationResult;
+        private final FailureFactory failureFactory;
+        private final boolean isConfigurationOnDemand;
+        private final boolean isOwnerBuildSuccessfullyConfigured;
+        private final String modelName;
+
+        public ResilientToolingModelBuilder(
+            ToolingModelBuilderLookup.Builder delegate,
+            ConfigurationResult result,
+            boolean isConfigurationOnDemand,
+            boolean isOwnerBuildSuccessfullyConfigured,
+            FailureFactory failureFactory,
+            String modelName
+        ) {
+            this.delegate = delegate;
+            this.configurationResult = result;
+            this.isConfigurationOnDemand = isConfigurationOnDemand;
+            this.isOwnerBuildSuccessfullyConfigured = isOwnerBuildSuccessfullyConfigured;
+            this.failureFactory = failureFactory;
+            this.modelName = modelName;
+        }
+
+        @Override
+        public @Nullable Class<?> getParameterType() {
+            return delegate.getParameterType();
+        }
+
+        @Override
+        public Object build(@Nullable Object parameter) {
+            if (canAssumeProjectFullyConfigured()) {
+                return delegate.build(parameter);
+            }
+
+            Object model = canRunEvenIfProjectNotFullyConfigured(modelName) ? delegate.build(parameter) : null;
+            List<Failure> failures = configurationExceptionAsFailure(configurationResult);
+            return ToolingModelBuilderResultInternal.attachFailures(model, failures);
+        }
+
+        private boolean canAssumeProjectFullyConfigured() {
+            if (configurationResult.isSuccess()) {
+                // If there was no configuration failure, then the project was configured successfully.
+                return true;
+            }
+
+            // For configure-on-demand, projects are configured individually, so failure at this point means that the project failed to configure.
+            // For normal mode assume project was successfully configured if the owner build was fully configured, e.g., full included build configuration was successful.
+            return !isConfigurationOnDemand && isOwnerBuildSuccessfullyConfigured;
+        }
+
+        private static boolean canRunEvenIfProjectNotFullyConfigured(String modelName) {
+            // Some internal model builders can run even if the project is not fully configured.
+            return RESILIENT_MODELS.contains(modelName);
```

Comment:
💭 We could expose actual builder type by adding a new to `ToolingModelBuilderLookup.Builder`. And then  `resilient model builders` could implement some new interface.

But I left this out for now to not have too many changes here.